SpockOpaleye/src/ApiBase.hs

```diff
@@ -0,0 +1,157 @@
+{-# LANGUAGE FlexibleContexts       #-}
+{-# LANGUAGE FlexibleInstances      #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE MultiParamTypeClasses  #-}
+{-# LANGUAGE OverloadedStrings      #-}
+{-# LANGUAGE TemplateHaskell        #-}
+
+module ApiBase where
+
+import           Control.Lens
+import           Control.Monad.IO.Class
+import           Control.Monad.Reader
+import           Control.Monad.Writer
+import qualified Data.Profunctor.Product.Default as D
+import           Data.Time                       (UTCTime, getCurrentTime)
+import           DataTypes
+import           Opaleye
+import           OpaleyeDef
+import qualified Data.Text as T
+import           GHC.Int
+import           Prelude                         hiding (id)
+import           TH
+import           Data.Aeson (Value(..))
+import           JsonInstances ()
+
+makeAudtableLenses ''Role
+makeAudtableLenses ''Tenant
+makeAudtableLenses ''User
+
+auditLog :: String -> AppM ()
+auditLog = tell 
+
+removeRawDbRows :: Table columnsW columnsR -> (columnsR -> Column PGBool) -> AppM GHC.Int.Int64
+removeRawDbRows table matchFunc = do
+  conn <- getConnection
+  liftIO $ runDelete conn table matchFunc 
+
+createDbRows :: (Show columnsW, D.Default QueryRunner columnsR haskells) 
+    =>  Table columnsW columnsR -> [columnsW] -> AppM [haskells]
+createDbRows table pgrows = do
+  auditLog $ "Create : " ++ (show pgrows)
+  conn <- getConnection
+  liftIO $ runInsertManyReturning conn table pgrows (\x -> x)
+
+updateDbRow :: (Show columnsW, HasId columnsR (Column PGInt4)) => Table columnsW columnsR -> Column PGInt4 -> columnsW -> AppM columnsW
+updateDbRow table row_id item = do
+  auditLog $ "Update :" ++ (show item)
+  conn <- getConnection
+  _ <- liftIO $ runUpdate conn table (\_ -> item) (matchFunc row_id) 
```

Comment:
Yes. That would be ideal. But since we already hold the record in its updated state, I thought, why do the additional fetching. If there was a failure with update, we will get an exception anyway, right?

I am not sure, but I think the use case of "update returning" would be if we are doing an operation on existing value, like "update product set quantity = quantity + 1", and in that case, returning the new values can save an additional select.

But here we are not doing anything like that.